server/blob/blob-cloud/src/main/java/org/apache/james/blob/cloud/CloudBlobsDAO.java

```diff
@@ -0,0 +1,120 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.cloud;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.HashBlobId;
+import org.apache.james.blob.api.ObjectStore;
+import org.apache.james.blob.api.ObjectStoreException;
+import org.jclouds.ContextBuilder;
+import org.jclouds.blobstore.BlobStore;
+import org.jclouds.blobstore.domain.Blob;
+import org.jclouds.blobstore.options.CopyOptions;
+import org.jclouds.logging.slf4j.config.SLF4JLoggingModule;
+import org.jclouds.openstack.swift.v1.blobstore.RegionScopedBlobStoreContext;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.hash.Hashing;
+import com.google.common.hash.HashingInputStream;
+import com.google.inject.Inject;
+import com.google.inject.Module;
+
+class CloudBlobsDAO implements ObjectStore {
+    private static final InputStream EMPTY_STREAM = new ByteArrayInputStream(new byte[0]);
+    private static final Iterable<Module> JCLOUDS_MODULES = ImmutableSet.of(new SLF4JLoggingModule());
+
+    private final BlobId.Factory blobIdFactory;
+    private final BlobStore blobStore;
+    private final ContainerName containerName;
+
+    @Inject
+    public CloudBlobsDAO(ContainerName containerName, HashBlobId.Factory blobIdFactory,
+                         CloudsBlobsConfiguration cloudsBlobsConfiguration) {
+        this.blobIdFactory = blobIdFactory;
+        this.containerName = containerName;
+
+        RegionScopedBlobStoreContext blobStoreContext = ContextBuilder.newBuilder("openstack-swift")
+            .endpoint(cloudsBlobsConfiguration.getEndpoint().toString())
+            .credentials(
+                cloudsBlobsConfiguration.getIdentity().getValue(),
+                cloudsBlobsConfiguration.getCredentials().getValue())
+            .overrides(cloudsBlobsConfiguration.getOverrides())
+            .modules(JCLOUDS_MODULES)
+            .buildView(RegionScopedBlobStoreContext.class);
+
+        blobStore = cloudsBlobsConfiguration
+            .getRegion()
+            .map(region -> blobStoreContext.getBlobStore(region.getValue()))
+            .orElse(blobStoreContext.getBlobStore());
+    }
+
+    @Override
+    public CompletableFuture<BlobId> save(byte[] data) {
+        return save(new ByteArrayInputStream(data));
+    }
+
+    @Override
+    public CompletableFuture<BlobId> save(InputStream data) {
```

Comment:
I think this method is a bit too long (or at least contains too many steps), what about : 
```
(tmpId, hash) = saveWithTmpId(data);
changeBlobId(tmpId, hash);
```

Reply:
I see what you mean, I'll look into it.
```
(tmpId, hash) = saveWithTmpId(data);
```
tuple decomposition in java ? now that's a new one ;)

Reply:
that's pseudo-code of course, you can rely on Pair class